File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-0.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def lowestCommonAncestor(self, root: TreeNode, p: TreeNode, q: TreeNode) -> TreeNode:
        if not root:
            return None
        if root.val == p.val or root.val == q.val:
            return root
        s, l = min(p.val, q.val), max(p.val, q.val)
        if root.val > s and root.val < l:
            return root
        if root.val > l:
            return self.lowestCommonAncestor(root.left, p, q)
        if root.val < s:
            return self.lowestCommonAncestor(root.right, p, q)
        

File: Data Structures & Algorithms/serialize-and-deserialize-binary-tree/submission-0.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Codec {
public:

    // Encodes a tree to a single string.
    string serialize(TreeNode* root) {
        string str = "";
        //TreeNode* p = root;
        queue<TreeNode*> q;
        q.push(root);
        while(!q.empty()) {
            auto top = q.front();
            q.pop();
            if(top) {
                str += to_string(top->val);
                q.push(top->left);
                q.push(top->right); 
            }
            else
                str += 'N';            
            str += ' ';                       
        }

        cout << str;
        return str;
    }

    // Decodes your encoded data to tree.
    TreeNode* deserialize(string data) {
        stringstream ss(data);
        TreeNode* root = NULL;
        string str = "";
        queue<TreeNode*> q;
        ss >> str;
        if(str == "N") {
            return root;
        }
        root = new TreeNode(stoi(str));
        q.push(root);

        while(!q.empty()) {
            auto top = q.front();
            q.pop();
            ss >> str;
            top->left = NULL;
            if(str != "N") {
                top->left =  new TreeNode(stoi(str));
                q.push(top->left);
            }
            ss >> str;
            top->right = NULL;
            if(str != "N") {
                top->right = new TreeNode(stoi(str));
                q.push(top->right);
            }
        }
        return root;
    }
};
